alf/algorithms/qrsac_algorithm.py

```diff
@@ -0,0 +1,189 @@
+            action_spec,
+            reward_spec=reward_spec,
+            actor_network_cls=actor_network_cls,
+            critic_network_cls=critic_network_cls,
+            epsilon_greedy=epsilon_greedy,
+            use_entropy_reward=use_entropy_reward,
+            normalize_entropy_reward=normalize_entropy_reward,
+            calculate_priority=calculate_priority,
+            num_critic_replicas=num_critic_replicas,
+            env=env,
+            config=config,
+            critic_loss_ctor=critic_loss_ctor,
+            target_entropy=target_entropy,
+            prior_actor_ctor=prior_actor_ctor,
+            initial_log_alpha=initial_log_alpha,
+            max_log_alpha=max_log_alpha,
+            target_update_tau=target_update_tau,
+            target_update_period=target_update_period,
+            dqda_clipping=dqda_clipping,
+            actor_optimizer=actor_optimizer,
+            critic_optimizer=critic_optimizer,
+            alpha_optimizer=alpha_optimizer,
+            debug_summaries=debug_summaries,
+            reproduce_locomotion=reproduce_locomotion,
+            name=name)
+
+        self._min_critic_by_critic_mean = min_critic_by_critic_mean
+        assert self._act_type == ActionType.Continuous, (
+            "Only continuous action space is supported for qrsac algorithm.")
+
+    def _compute_critics(self,
+                         critic_net,
+                         observation,
+                         action,
+                         critics_state,
+                         replica_min: bool = True,
+                         quantile_mean: bool = True):
+        critic_inputs = (observation, action)
+        critic_quantiles, critics_state = critic_net(
+            critic_inputs, state=critics_state)
+
+        # For multi-dim reward, do:
+        #   [B, replicas * reward_dim] -> [B, replicas, reward_dim]
+        # For scalar reward, do nothing
+        if self.has_multidim_reward():
+            remaining_shape = critic_quantiles.shape[2:]
+            critic_quantiles = critic_quantiles.reshape(
+                -1, self._num_critic_replicas, *self._reward_spec.shape,
+                *remaining_shape)
+        if replica_min:
+            if self._min_critic_by_critic_mean:
+                # [B, replicas] or [B, replicas, reward_dim]
+                critic_mean = critic_quantiles.mean(-1)
+                idx = torch.min(
+                    critic_mean, dim=1)[1]  # [B] or [B, reward_dim]
+                if self.has_multidim_reward():
+                    idx = torch.repeat_interleave(
+                        idx.unsqueeze(-1), critic_quantiles.shape[-1],
+                        dim=-1)  # [B, reward_dim, n_quantiles]
+                    idx = idx.unsqueeze(1)  # [B, 1, reward_dim, n_quantiles]
+                    # [B, reward_dim, n_quantiles]
+                    critic_quantiles = torch.gather(
+                        critic_quantiles, dim=1, index=idx).squeeze()
```

Comment:
If familiar with array indexing (https://numpy.org/doc/stable/reference/arrays.indexing.html):
the following seems more understandable:
```python
B, replicas, reward_dim = critic_mean.shape
critic_quantiles = critic_quantiles[torch.arange(B)[:, None], idx, torch.arange(reward_dim)]
```

Reply:
I don't quite get this, let's discuss it offline.